deepchem/splits/__init__.py

```diff
@@ -5,188 +5,291 @@
+
+    def train_test_split(self, samples, train_dir, test_dir, seed=None,
+                         frac_train=.8):
+        """
     Splits self into train/test sets.
 
     Returns Dataset objects.
     """
-    valid_dir = None
-    train_samples, _, test_samples = self.train_valid_test_split(
-        samples, train_dir, valid_dir, test_dir,
-        frac_train=frac_train, frac_test=1-frac_train, frac_valid=0.)
-    return train_samples, test_samples
-
-  def split(self, samples, frac_train=None, frac_valid=None, frac_test=None,
-            log_every_n=None):
-    """
+        valid_dir = None
+        train_samples, _, test_samples = self.train_valid_test_split(
+            samples, train_dir, valid_dir, test_dir,
+            frac_train=frac_train, frac_test=1 - frac_train, frac_valid=0.)
+        return train_samples, test_samples
+
+    def split(self, samples, frac_train=None, frac_valid=None, frac_test=None,
+              log_every_n=None):
+        """
     Stub to be filled in by child classes.
     """
-    raise NotImplementedError
+        raise NotImplementedError
 
-class MolecularWeightSplitter(Splitter):
+class StratifiedSplitter(Splitter):
+    """
+  Class for doing stratified splits -- where data is too sparse to do regular splits
   """
+
+    def __randomize_arrays(self, array_list):
+      #assumes that every array is of the same dimension
+      numRows = array_list[0].shape[0]
+      perm = np.random.permutation(numRows)
+      for array in array_list:
+          array = array[perm]
+      return array_list
+
+    def __generate_required_hits(self, w, frac_split):
+      required_hits = (w != 0).sum(0)#returns list of per column sum of non zero elements
+      for colHits in required_hits:
+        colHits = int(frac_split * colHits)
+      return required_hits
+
+    def __generate_required_index(self, w, required_hit_list):
+      colIndex = 0
+      index_hits = []
+      #loop through each column and obtain index required to splice out for required fraction of hits
+      for col in w.T:
+          num_hit = 0
+          num_required = required_hit_list[colIndex]
+          for index, value in enumerate(col):
+              if value != 0:
+                num_hit +=1
+                if num_hit >= num_required:
+                  index_hits.append(index)
+                  break
+          colIndex += 1
+      return index_hits
+
+    def __split(self, X, y, w, ids, frac_split):
+      """
+      Method that does bulk of splitting dataset appropriately based on desired split percentage
+      """
+      #find he total number of hits for each task and calculate the required
+      #number of hits for split based on frac_split
+      required_hits_list = self.__generate_required_hits(w, frac_split)
+      #finds index cutoff per task in array to get required split calculated
+      index_list = self.__generate_required_index(w, required_hits_list)
+
+      X_first_split = X_second_split = X
```

Comment:
Using short variable names could make this function more readable. Simply `X_1, X_2` instead of `X_first_split`, `X_second_split` for example.